ksql-engine/src/main/java/io/confluent/ksql/engine/Checksum.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.engine;
+
+import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * This class represents data that provides enough information to identify
+ * whether two instances of {@link io.confluent.ksql.KsqlExecutionContext}
+ * are equivalent.
+ *
+ * <p>Today, it is represented as a chain of hashes of all statements that
+ * were successfully executed. Any other information that may cause engines
+ * to behave differently should be added to this checksum.</p>
+ */
+@JsonIgnoreProperties(ignoreUnknown = true)
+public final class Checksum {
```

Comment:
I will definitely do this, though if there were threading issues beforehand there would still be threading issues now because we have this pattern:
```java
    final ExecuteResult result = executor.execute(statement);
    checksum = checksum.update(statement);
    return result;
```
If those are not atomic (and concurrent operations happen), it is still possible that two threads update using the old checksum value 😲! At least it makes sure that `Checksum` itself is thread safe.

Perhaps the more fundamental question is whether or not we have concurrent query executions! I will answer that question, and if so I will make these operations atomic.